ambry-store/src/test/java/com.github.ambry.store/CuratedLogIndexState.java

```diff
@@ -393,6 +474,119 @@ void verifyEntriesForHardDeletes(Set<MockId> deletedIds) throws IOException {
+    return deleteIndexSegmentStartOffset != null
+        && lastModifiedTimesInSecs.get(deleteIndexSegmentStartOffset) * Time.MsPerSec < referenceTimeMs;
+  }
+
+  /**
+   * Returns {@code true} if the {@code id} is considered expired at {@code referenceTimeMs}.
+   * @param id the {@link MockId} to check.
+   * @param referenceTimeMs the reference time in ms at which the check needs to be performed.
+   * @return {@code true} if the {@code id} is considered expired at {@code referenceTimeMs}. {@code false} otherwise.
+   */
+  boolean isExpiredAt(MockId id, long referenceTimeMs) {
+    long expiresAtMs = allKeys.get(id).getFirst().getExpiresAtMs();
+    return expiresAtMs != Utils.Infinite_Time && expiresAtMs < referenceTimeMs;
+  }
+
+  /**
+   * Ensures that the {@link PersistentIndex} is sane and correct by checking that
+   * 1. It contains no duplicate entries.
+   * 2. The ordering of PUT and DELETE entries is correct.
+   * 3. There are no offsets in the log not accounted for in the index.
+   * @throws IOException
+   * @throws StoreException
+   */
+  void verifyRealIndexSanity() throws IOException, StoreException {
+    Map<MockId, Boolean> keyToDeleteSeenMap = new HashMap<>();
+    IndexSegment prevIndexSegment = null;
+    for (IndexSegment indexSegment : index.getIndexSegments().values()) {
+      Offset indexSegmentStartOffset = indexSegment.getStartOffset();
+      if (prevIndexSegment == null) {
+        assertEquals("There are offsets in the log not accounted for in index", log.getFirstSegment().getStartOffset(),
+            indexSegmentStartOffset.getOffset());
+      } else if (prevIndexSegment.getLogSegmentName().equals(indexSegment.getLogSegmentName())) {
+        assertEquals("There are offsets in the log not accounted for in index", prevIndexSegment.getEndOffset(),
+            indexSegmentStartOffset);
+      } else {
+        LogSegment segment = log.getSegment(prevIndexSegment.getLogSegmentName());
+        assertEquals("There are offsets in the log not accounted for in index", segment.getEndOffset(),
+            prevIndexSegment.getEndOffset().getOffset());
+        segment = log.getNextSegment(segment);
+        assertEquals("There are offsets in the log not accounted for in index", segment.getStartOffset(),
+            indexSegmentStartOffset.getOffset());
+      }
+      NavigableSet<IndexEntry> indexEntries = new TreeSet<>(PersistentIndex.INDEX_ENTRIES_COMPARATOR);
+      List<MessageInfo> infos = new ArrayList<>();
+      FindEntriesCondition condition = new FindEntriesCondition(Long.MAX_VALUE);
+      indexSegment.getEntriesSince(null, condition, infos, new AtomicLong(0));
+
+      for (MessageInfo info : infos) {
+        MockId id = (MockId) info.getStoreKey();
+        IndexValue value = indexSegment.find(id);
+        indexEntries.add(new IndexEntry(id, value));
+        Boolean deleteSeen = keyToDeleteSeenMap.get(id);
+        if (info.isDeleted()) {
+          if (deleteSeen != null) {
```

Comment:
For the first comment,
Yes I use it with the `BlobStoreCompactor` (easier to keep that PR focussed and tag any changes for test with this one). Could you describe how what you suggested is different from the code written? The code written basically calls `assertFalse` on `deleteSeen` if it is not `null`.
Will talk to you about the second comment - I'm not sure I understand what the scenario you are talking about is.